Documentation: stepsize and Newton Solver (#777)
festim-dev/FESTIM: docs/source/userguide/newton_solver.rst

=============
Newton solver
=============

> For advanced simulations, the parameters of the Newton solver can be adopted depending on a specific problem. 

Reviewer: ```suggestion
For advanced simulations, the parameters of the Newton solver can be adapted depending on a specific problem. 
```

-----------------
Built-in options
-----------------
A limited set of the solver parameters can be accessed via the built-in attributes of classes. 

> The parameters of the Newton solver for :class:`festim.HTransportProblem` can be chosen in :class:`festim.Settings` (see :ref:`settings_ug` for an example). Absolute and relative tolerances of the Newton solver

Reviewer: ```suggestion
The parameters of the Newton solver for :class:`festim.HTransportProblem` can be chosen in :class:`festim.Settings` (see :ref:`settings_ug`). Absolute and relative tolerances of the Newton solver
```

are defined with ``absolute_tolerance`` and ``relative_tolerance`` attributes, respectively. The maximum number of the solver iterations can be set using 

> the ``maximum_iterations`` parameter. Additionally, there is an option to choose linear solver and preconditioning methods that suit better for a particular problem.

Reviewer: ```suggestion
the ``maximum_iterations`` parameter. Additionally, there is an option to choose linear solver and preconditioning methods that may be more suitable for particular problems.
```

The linear solver method can be set with the ``linear_solver`` attribute. The list of available linear solvers can be viewed with: ``print(fenics.list_linear_solver_methods())``.

> .. dropdown:: Linear solver methods

Reviewer: Nice, didn't know that 👍

    * "mumps" - MUMPS (MUltifrontal Massively Parallel Sparse direct Solver)
    * "petsc" - PETSc built in LU solver
    * "richardson" - Richardson method 
    * "superlu" - SuperLU
    * "superlu_dist" - Parallel SuperLU
    * "tfqmr" - Transpose-free quasi-minimal residual method
    * "umfpack" - UMFPACK (Unsymmetric MultiFrontal sparse LU factorization)

The preconditioner can be set with the ``preconditioner`` attribute. The list of available preconditioners can be viewed with: ``print(fenics.list_krylov_solver_preconditioners())``.

.. dropdown:: Preconditioners

    * "amg" - Algebraic multigrid
    * "hypre_amg" - Hypre algebraic multigrid (BoomerAMG)
    * "hypre_euclid" - Hypre parallel incomplete LU factorization
    * "hypre_parasails" - Hypre parallel sparse approximate inverse
    * "icc" - Incomplete Cholesky factorization
    * "ilu" - Incomplete LU factorization
    * "jacobi" - Jacobi iteration 
    * "petsc_amg" - PETSc algebraic multigrid
    * "sor" - Successive over-relaxation

Similarly, the Newton solver parameters of :class:`festim.HeatTransferProblem`, :class:`festim.ExtrinsicTrap`, or :class:`festim.NeutronInducedTrap` 

> can be defined upon the need. Here is an example for the heat transfer problem:

Reviewer: ```suggestion
can be defined if needed. Here is an example for the heat transfer problem:
```

.. code-block:: python

    model.T = HeatTransferProblem(
        transient=True,
        initial_condition=300,
        absolute_tolerance=1.0,
        relative_tolerance=1e-10,
        maximum_iterations=50,
        linear_solver="gmres",
        preconditioner="icc")

--------------
Custom solver
--------------

> For a more flexible control, the built-in Newton solver can be overwritten by a custom solver based on the ``fenics.NewtonSolver`` class. 

Reviewer: ```suggestion
For a finer control, the built-in Newton solver can be overwritten by a custom solver based on the :class:`festim.NewtonSolver` class. 
```

Author: Why `festim.NewtonSolver`?

Reviewer: Oh yes sorry my bad!

A user-defined Newton solver can be provided after :class:`festim.Simulation.initialise()`:

.. code-block:: python

>     class CustomSolver(f.NewtonSolver):
>         def __init__(self):
>             super().__init__()
> 
>     custom_solver = CustomSolver()

Reviewer: Since you're not modifying any of the class functionality you can simply do this no?

```suggestion
    import fenics as f

    custom_solver = f.NewtonSolver()
```

It may be worth showing an example with a modification of the functionality too then

Author: I can add both examples: with/without modifications.

What do you think of [this](https://fenicsproject.discourse.group/t/set-krylov-linear-solver-paramters-in-newton-solver/1070/3) example with a modification of the functionality?

Reviewer: Or maybe the one that you use for convergence plots?

Author: Yeah. Will add.

For this initial attempt, I just didn't want to write something complicated, but rather illustrate the possibility of using the overwritten `fenics.NewtonSolver`.

Reviewer: Up to you Vladimir!

    custom_solver.parameters["error_on_nonconvergence"] = False
    custom_solver.parameters["absolute_tolerance"] = 1e10
    custom_solver.parameters["relative_tolerance"] = 1e-10
    custom_solver.parameters["maximum_iterations"] = 100
    custom_solver.parameters["linear_solver"] = "gmres"
    custom_solver.parameters["preconditioner"] = "ilu"

    model.initialise()

    model.h_transport_problem.newton_solver = custom_solver

    model.run()

.. warning::
    

>     For a stationary heat transfer problem, a custom Newton solver has to be provided before the simulation initialisation! 

Reviewer: Could we add another warning saying that this will override the solver parameters given in `Settings`?
